src/fifobuffer.jl

```diff
@@ -79,10 +80,10 @@ Base.read(f::FIFOBuffer) = readavailable(f)
 Base.flush(f::FIFOBuffer) = nothing
 
 function Base.eof(f::FIFOBuffer)
-    return f.eof && f.nb == 0
+    return f.content_length >= 0 && f.content_pos >= f.content_length || f.content_length < 0 && f.nb == 0
 end
 function Base.close(f::FIFOBuffer)
-    f.eof = true
+    f.content_length = f.content_pos
```

Comment:
> I think this case is a little more problematic; what if the buffer gets closed before `content_pos` has reached `content_length`?

If a user attempts to read the last available byte in the buffer then the read will be blocked. It is up to the buffer writer to either close the `FIFOBuffer` which could potentially stop a read before it finished reading all available data or update the `content_length` such that reading automatically stops.

> I guess it somewhat depends on if `content_pos` is tracking the # of bytes *written* to the buffer, or # of bytes *read* from buffer.

`content_pos` is definitely tracking the number of bytes *read* from the buffer. It may make sense to rename these `read_pos` instead of `content_pos` and `expected_length` instead of `content_length`.